### ISSUE TITLE
PMOsc implementation missing .mod(2pi)

### ISSUE THREAD

**Issue by elifieldsteel** [supercollider/supercollider#4408]
## Environment

* SuperCollider version: 3.10.2
* Operating system: macOS 10.11.6
* Other details (Qt version, audio driver, etc.): N/A

## Steps to reproduce

```supercollider
Paste code here. Use a minimal example please.

s.boot;

//PMOsc produces results outside the boundaries of "classic" FM/PM synthesis when
//pmindex > 8pi, creating a "staircasing" pitch effect. This is because the modulating
//SinOsc does not include .mod(2pi) in the implementation code.
{PMOsc.ar(500, 4, 40) * 0.2!2}.play;

//should sound like this
{SinOsc.ar(500, SinOsc.ar(4, mul:40).mod(2pi)) * 0.2!2}.play;


//The discrepancy disappears when pmindex < 8pi:
{PMOsc.ar(500, 4, 10) * 0.2!2}.play;
//vs
{SinOsc.ar(500, SinOsc.ar(4, mul:10).mod(2pi)) * 0.2!2}.play;

//PMOsc implementation should probably look like this, with .mod(2pi)
//added to both .ar and.kr methods:

PMOsc  {

	*ar { arg carfreq,modfreq,pmindex=0.0,modphase=0.0,mul=1.0,add=0.0;
		^SinOsc.ar(carfreq, SinOsc.ar(modfreq, modphase, pmindex).mod(2pi),mul,add)
	}

	*kr { arg carfreq,modfreq,pmindex=0.0,modphase=0.0,mul=1.0,add=0.0;
		^SinOsc.kr(carfreq, SinOsc.kr(modfreq, modphase, pmindex).mod(2pi),mul,add)
	}

}
```

## Expected vs. actual behavior

See above.

**Comment by kureta** [comment]
I'm completely new to supercollider but wouldn't it be better, in terms of performance, to apply `.mod(2pi)` to `pmindex` instead? This way if you provide a constant value as the `pmindex` argument, the calculation is done only once, whereas in the suggested version it seems like the calculation will be done 44100 times per second (or whatever the current sample rate is). To clarify:

    SinOsc.ar(carfreq, SinOsc.ar(modfreq, modphase, pmindex.mod(2pi))

Again, I don't know anything about the inner workings of supercollider, I'm pointing this out mostly to learn.

**Comment by elifieldsteel** [comment]
No, although I can see the logic behind your suggestion, this is incorrect. pmindex is the mul input of the inner SinOsc. By wrapping it between [0, 2pi], you're limiting the modulation index that the user can specify, and therefore limiting the complexity of spectrum the user can produce. The pmindex input should be allowed to be arbitrarily large. It is specifically the _phase_ values for SinOsc that must be wrapped between [0, 2pi], so .mod(2pi) should be applied to the inner SinOsc. Note the difference in sound between these two lines of code:

truncates the modulation index to ~6.01, but the user wants an index of 50:
`{SinOsc.ar(500, SinOsc.ar(4, mul:50.mod(2pi))) * 0.2!2}.play;`

correctly wraps the sinusoidal modulator within appropriate bounds:
`{SinOsc.ar(500, SinOsc.ar(4, mul:50).mod(2pi)) * 0.2!2}.play;`

I suppose the modphase argument also needs to be constrained between [0, 2pi], but the user has total control over this value, so it's their responsibility/decision whether they want to wrap modphase.

Side note, doing a calculation like .mod(2pi) at the sampling rate is dirt cheap—it's something that scsynth is specifically designed to do, and it's comfortable doing a ton of calculations like this.

**Comment by kureta** [comment]
@elifieldsteel thank you very much for the detailed response. I actually came here straight from your latest tutorial on YouTube. Also thank you for those awesome tutorials!
